### PR TITLE
Add `Span` struct (replacing `StrRange`). Spans represent read-only access to a contiguous array, resembling `std::span`.

### PR DESCRIPTION
**Pull request by Ivorforce**
It is currently difficult to run algorithms on `String`, `Vector`, `LocalVector`  data ranges or plain C arrays without copying the data first. This leads to inefficiencies.

This PR adds the `Span` class. A span represents a view into an array (it's a `pointer` and `size`). 
With `Span`, it will be easier to implement functions with more agnosticism as to the memory storage, helping to reduce unnecessary copies. Additionally, `Span` will help bridge the gap between `LocalVector` and `Vector`, helping to address https://github.com/godotengine/godot-proposals/issues/5144.

`Span` is similar to `StrView` which it replaces, but meant for more than just strings. It is furthermore similar to `VectorView` which will be replaced in a future PR.

## Example
The `String.path_to` implementation currently has the following lines of code:

https://github.com/godotengine/godot/blob/c2e4ae782a399adf42c664d3d21d8defe49ba8e5/core/string/ustring.cpp#L5108-L5109

Here, the String is first subselected through `substr`, causing a copy to be made.
Then, it is split through a `split` call, through which multiple copies of regions of the string are made.
None of the regions are modified in the process - the copies are made merely because it is not possible to avoid them.

An optimized implementation could look like this:
```c++
LocalVector<Span<char32_t>> src_dirs = spans::split(src.span().subspan(1, src.length() - 2), U'/'); 
LocalVector<Span<char32_t>> dst_dirs = spans::split(dst.span().subspan(1, dst.length() - 2), U'/'); 
```

In this implementation, no copies of the string need to be made, because all `Span`s used here are views into the original string (`spans::split` and `subspan` would need to be added as functions).

## Discussion
`Span` is named after `std::span` of C++20 [std::span](https://en.cppreference.com/w/cpp/container/span).
`Span` is const-only because a need for mutable spans is not obvious yet. It is easier to use if it is const by default. If mutable spans are needed in the future, it can be proposed then.

### CODE REVIEW

**Comment by Ivorforce** [comment]
Looks like the builds are failing.
The Windows failure is a repeat of a problem I had in #99806 already. It can be fixed by touching the file to wipe caches.
I have no idea what the Linux problem is about though, since tests are completing on my own machine.

**Comment by RandomShaper** [comment]
There's a `VectorView` somewhere in the rendering code (in `rendering_device_commons.h`, I think) that seems to have the same goal. Also, I barely remember there's a proposal around related to this.

**Comment by Ivorforce** [comment]
> There's a `VectorView` somewhere in the rendering code (in `rendering_device_commons.h`, I think) that seems to have the same goal. Also, I barely remember there's a proposal around related to this.

Oh yeah, [I see it](https://github.com/godotengine/godot/blob/master/servers/rendering/rendering_device_driver.h#L60).
I think it would be best to consolidate that into `BufferView` once merged, in a follow-up PR.

**Comment by Ivorforce** [comment]
I am touching `feed_effects.h` by inserting an explicit include to a well-known file. 
I think MSVC cache error mentioned above may be caused by the only include of the file being a generated file, possibly leading to the cache resolver running into a race condition when compiling the file. I hope this change can fix the problem for good (though we won't know for sure until future PRs since touching the file already invalidates the cache, fixing the issue).

**Comment by Ivorforce** [comment]
I renamed `BufferView` -> `Span` since that's what C++20 calls it. I also changed the implementation around a tiny bit to match what may be expected of the type.

File: core/templates/span.h [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+class Span {
+	T *_ptr = nullptr;
+	size_t _len = 0;
+
+public:
+	Span() = default;
+	Span(T *p_ptr, size_t p_len) :
+			_ptr(p_ptr), _len(p_len) {}
+
+	T *ptrw() { return _ptr; }
```
[the remark this answers is not in the document]
Comment (hpvb) [reply core/templates/span.h:48]: Actually no, it wouldn't, not directly, as vieww() would already call _ref(). I think then that we need to make default, copy, and move assignment and constructors private of this thing so it can't escape scope. I have some concerns about these things ending up outliving insertions or removals from cowdata.

[the remark this answers is not in the document]
Comment (Ivorforce) [reply core/templates/span.h:48]: It doesn't, but it _will_ behave unexpectedly if trying to access a `ptrw` from a const view:

```
Span<const int> span = _cowdata.view();
const int *ptr = span.ptrw();  // Since T is const int
```

I think I better use SFINAE to disable `ptrw` for `is_const_v<T>` types.

[the remark this answers is not in the document]
Comment (Ivorforce) [reply core/templates/span.h:48]: It's a bit complex, but I think this is better: 
```c++
std::enable_if<std::negation_v<std::is_const<T>>, std::remove_const_t<T>> *ptrw() { return _ptr; }
```

[the remark this answers is not in the document]
Comment (hpvb) [reply core/templates/span.h:48]: That would be good I think. Too bad there's not really a way to ensure that something like 

```
y = x.view()
x.insert(blah)
y.find("something")
```

can't happen. Or worse that y gets passed to something else and stored.

[the remark this answers is not in the document]
Comment (Ivorforce) [reply core/templates/span.h:48]: Yeah, that's going to be an inherent risk with this data type. Would be possible to avoid in Rust, but not good old C++

[the remark this answers is not in the document]
Comment (Ivorforce) [reply core/templates/span.h:48]: Yeah, makes sense. I'll add a comment about the SFINAE use. I will also add a disclaimer to not store this data type.

[the remark this answers is not in the document]
Comment (hpvb) [reply core/templates/span.h:48]: might help to delete copy constructors and such, with a comment as to why. at least it will catch some cases of likely abuse.

[the remark this answers is not in the document]
Comment (hpvb) [reply core/templates/span.h:48]: Okay, let's not do that then. It's no more dangerous than sending a ptr() or ptrw() anyway, so maybe it's not actually a problem

File: core/templates/span.h [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+class Span {
+	T *_ptr = nullptr;
+	size_t _len = 0;
+
+public:
+	static constexpr bool is_mutable = std::negation_v<std::is_const<T>>;
+
+	Span() = default;
+	Span(T *p_ptr, size_t p_len) :
+			_ptr(p_ptr), _len(p_len) {}
```
Comment (Repiteo) [inline core/templates/span.h:51]: If you wanna get *really* fancy, you can make the length argument implicit:
```suggestion
	Span() = default;
	Span(T *p_ptr, size_t p_len) :
			_ptr(p_ptr), _len(p_len) {}
	template <size_t len>
	Span(T (&p_ptr)[len]) :
			_ptr(p_ptr), _len(len) {}
```

**Review by Repiteo — COMMENTED** [review]
I'd encourage adding tests as well. Seeing as you're making a new template, you'd be free to make several functions `constexpr` for compile-time sanity checks:
```cpp
TEST_CASE("[Span] Constant Validators") {
	constexpr Span<uint8_t> span_empty;
	static_assert(span_empty.size() == 0);
	static_assert(span_empty.is_empty());

	constexpr uint8_t byte = 0;
	constexpr Span<uint8_t> span_byte = Span<uint8_t>(byte, 1);
	static_assert(span_byte.size() == 1);
	static_assert(!span_byte.is_empty());
}
```

**Comment by Ivorforce** [comment]
> I'd encourage adding tests as well. Seeing as you're making a new template, you'd be free to make several functions `constexpr` for compile-time sanity checks:
> 
> ```c++
> TEST_CASE("[Span] Constant Validators") {
> 	constexpr Span<uint8_t> span_empty;
> 	static_assert(span_empty.size() == 0);
> 	static_assert(span_empty.is_empty());
> 
> 	constexpr uint8_t byte = 0;
> 	constexpr Span<uint8_t> span_byte = Span<uint8_t>(byte, 1);
> 	static_assert(span_byte.size() == 1);
> 	static_assert(!span_byte.is_empty());
> }
> ```

I like this test! I'll add some more as well.
I'm realizing as well, i can make _every single_ function currently in `Span` constexpr. Not just great for the compiler, but it also means we don't need a single runtime test 😄

File: core/templates/span.h (hunk 2 above)
Comment (Ivorforce) [reply core/templates/span.h:51]: This would unfortunately be incorrect for string types, since the `NULL` terminator is part of the `[len]`. I may add a follow up PR to add array literal constructors, and specialize string literal constructors to use `strlen()`, but I thought it too controversial to end up in the initial PR.

**Comment by Ivorforce** [comment]
Hrm... Wondering again if it wouldn't be possible to just have `const Span<char>` propagate constness to the contained type. It would just require a few const cast hacks. I'll have a look later and see if I can make it work.

**Comment by Ivorforce** [comment]
I've updated the PR according to feedback from the godot core meeting:
- `Span` will hold only `const` pointers for now. If mutable pointers are needed, they should be separately proposed.
- `operator[]` subscripts will do a sanity bounds check.
- `find`, `rfind` and `count` will not be hosted by `Span`. Instead, they should be hosted by a separate header (to be moved in another PR).

`VectorView` will be replaced with `Span` in a future PR.

The PR is ready for review again.

File: core/templates/span.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+// Equivalent of std::span.
+// Represents a view into a contiguous memory space.
+// DISCLAIMER: This data type does not own the underlying buffer. DO NOT STORE IT.
+//  Additionally, for the lifetime of the Span, do not resize the buffer, and do not insert or remove elements from it.
+//  Failure to respect this may lead to crashes or undefined behavior.
+template <typename T>
+class Span {
+	const T *_ptr = nullptr;
+	uint32_t _len = 0;
```
Comment (kiroxas) [inline core/templates/span.h:44]:
```
uint64_t _len = 0;
const T *_ptr = nullptr;
```
to have a memory layout similar to vector ( @RandomShaper  talked about it during the meeting)

**Review by kiroxas — APPROVED** [review]

File: core/templates/span.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ivorforce) [reply core/templates/span.h:44]: On 32-bit systems the current implementation should be size 64 pad 32, and on 64-bit systems it should be size 96 pad 64, no?

Comment (Ivorforce) [reply core/templates/span.h:44]: `Vector` doesn't store size locally (it stores it before the buffer itself), and `LocalVector` uses `count, capacity, pointer`. It's currently impossible match either of them 😅

Comment (kiroxas) [reply core/templates/span.h:44]: Not that it really matters anyway, but that align more closely to the CowData layout, which is the one use to talk with GdScript. But you can't match them both yeah :). (And I don't think directly casting raw memory of vector to span is a trick that will be used anyway)

[the remark this answers is not in the document]
Comment (kiroxas) [reply core/templates/span.h:44]: on 32 bit, you're just 64, but I'm not sure 4.x is compiled to 32 bits ? And on 64 you have 64bits of pointer, 32 bits of integer, and then you need to pad to be aligned to the max alignment of your fields (which is 64 due to the pointer)

[the remark this answers is not in the document]
Comment (Ivorforce) [reply core/templates/span.h:44]: >  then you need to pad to be aligned to the max alignment of your fields (which is 64 due to the pointer)

Padding comes before the struct, not after. If you have a `Span` as a variable, then a `uint32_t` as the next one, it can fit right after `_len` because it aligns to the 32 bit left by `Span`.

Comment (Ivorforce) [reply core/templates/span.h:44]: It doesn't match `CowData` because `CowData` is just a pointer, and the `size` is at the pointers' dereference:

```
ptr --> &data[0]

refcount
size
data....
```

It is impossible to design `Span` to match this because `CowData` does not contiguously store `ptr` and `size` at all.

[the remark this answers is not in the document]
Comment (kiroxas) [reply core/templates/span.h:44]: No, that's not right. Padding is at the end of the structure (or between the elements) **never** at the start. And if you declare a Span on the stack, it will pad, the next uin32_t variable will not be directly after your _len field. Except if you `#pragma pack(push, 1)` so that the compiler does not put padding bytes, but this is not the case here.

Comment (kiroxas) [reply core/templates/span.h:44]: You're right, that will not be useful. So any will do.

[the remark this answers is not in the document]
Comment (Ivorforce) [reply core/templates/span.h:44]: You're right! I had to [test it in a compiler](https://onlinegdb.com/cgZnOe0nK). I was operating under false assumptions here.
I suppose that means we could make it `uint64_t`, or just revert to `size_t` as I had it before to speed up 32 bit operation. What do you think?

[the remark this answers is not in the document]
Comment (YYF233333) [reply core/templates/span.h:44]: > Padding comes before the struct, not after. If you have a `Span` as a variable, then a `uint32_t` as the next one, it can fit right after `_len` without padding because it aligns to the 32 bit left by `Span`. 

I guess you are talking about this?

[the remark this answers is not in the document]
Comment (kiroxas) [reply core/templates/span.h:44]: I like `uint64_t` better **BUT** `size_t` is always packed (64 bits on x86 and 128 bits in x64) so ... if 4.x compiles for x62 32 bits, size_t makes some more sense

[the remark this answers is not in the document]
Comment (Ivorforce) [reply core/templates/span.h:44]: Yeah I guess `size_t` makes more sense. 64 bits difference can be crucial on 32 bit systems, especially if used as a pass-by-value struct.
I'm pushing that as a change, but leaving the conversation open for others to chip in.

File: servers/rendering/renderer_rd/effects/sort_effects.h [hunk 4]
```diff
@@ -31,6 +31,7 @@
 #ifndef SORT_EFFECTS_RD_H
 #define SORT_EFFECTS_RD_H
 
+#include "servers/rendering/renderer_rd/shader_rd.h"
```
Comment (Ivorforce) [inline servers/rendering/renderer_rd/effects/sort_effects.h:34]: Quick fix due to [build failing](https://github.com/godotengine/godot/actions/runs/12693042092/job/35379786940?pr=101352) because of erroneous cache restores.

I have https://github.com/godotengine/godot/pull/100293#issuecomment-2556731863 that including a non-generated file explicitly solves such cache issues for good.

File: core/templates/span.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (lawnjelly) [reply core/templates/span.h]: It might be worth some further discussion whether to use `size_t`.

We haven't tended to use it historically in containers - it seems like one of those vague flaky C types (like `long` etc) but most programmers these days have moved to fixed width types simply because vague types are a potential nightmare for multiplatform / compiler.

I understand the reasoning (particularly for keeping `Span` aligned), but if we want a `u32` or `u64` (depending on the target) I suspect we should consider specify this directly with a godot specific type. I'm not sure it would be a good idea to start riddling the codebase with `size_t` (which people might start doing as a result).

[the remark this answers is not in the document]
Comment (lawnjelly) [reply core/templates/span.h]: Another problem to think about in advance here by the way is that we might have a caller that uses something like:
```
for (u32 n=0; n<span.size(); n++)
```
This will end up happening, this is partly why variable size types are such a nightmare.

So it will afaik end up promoting n to 64 bit each time in e.g. a tight loop, just to make this comparison.

[the remark this answers is not in the document]
Comment (lawnjelly) [reply core/templates/span.h]: For these reasons actually I would reconsider fixing `size` as `u32` (similar to `LocalVector`), and using padding for 64 bit, as it will make usage far easier to reason about and remove platform effects.

In 99.9% of cases, `u32` gives 4gb range for 8 bit values, and more when addressing larger `T` sizes. For those cases that require more than that, they could use e.g. a 64 bit span?

Just my personal thoughts though.

[the remark this answers is not in the document]
Comment (kiroxas) [reply core/templates/span.h]: I'm on your side, as I tend to prefer fixed size to make the layout clearer and not platform dependant. This i why I wrote that :

> I like uint64_t better BUT ...

And size_t seems to be used a lot in the codebase anyway. 

> So it will afaik end up promoting n to 64 bit each time in e.g. a tight loop, just to make this comparison. (This might be free, or might not..)

When I look at assembly for this (and take this with a grain of salt, as some different usage _may_ cause different behaviour), and at my surprise, it is totally **not** free, as it forces one additional move ( [godbolt ](https://godbolt.org/)). It makes sense when you think about it, as it have to do the inc on a 32 bit register for overflow reasons, then move it to a 64 bits register for the comparison. Not a big deal, but it has a cost, so we should take it into consideration.

[the remark this answers is not in the document]
Comment (lawnjelly) [reply core/templates/span.h]: > But size_t seems to be used a lot in the codebase already....

`size_t` is (correctly) used where OS and API calls return a `size_t`, however historically we haven't (to my knowledge) used it much _internally_, for containers etc. But bear in mind I'm more familiar with 3.x.

I can see it used in `lru.h` (which may be new), but we should consider whether it should be used there.

[the remark this answers is not in the document]
Comment (Ivorforce) [reply core/templates/span.h]: Let me summarize my current thoughts.
`32 bit` sizes would put us at a disadvantage because 4gb buffers (at 1 byte objects) would not be supported. Such buffers are admittedly rare, but do happen.
`64 bit` sizes would be perfect on 64 bit systems, but slightly slow and potentially unsafe on 32 bit systems.
`size_t` is somewhat fishy because it's different sizes on different systems.

I'm still on the `size_t` side because it's fast, safe, and while we have no _guarantees_ C++ side, we do know it to be correct for all major platforms we expect Godot to ship on. It's also used by all C++ container types (`sizeof(array)`, `std::span`, `std::array`...).

**Review by Repiteo — APPROVED** [review]
Codestyle checks out & the use of more modern C++ concepts is a welcome addition

**Review by clayjohn — COMMENTED** [review]
I think this is ready to go. But we need to settle the `size_t` debate. Personally I lean towards just making it a `uint64_t`  and calling it a day. 

I appreciate the reasons for `size_t`, but I think the pain of having a changing size isn't going to be worth the slight optimization for 32 bit systems. If you look at download numbers, almost no one uses the 32 bit version of Godot. And, ultimately, we need to prioritize developer ergonomics in this case. Have a fixed size that we can count on and that matches our equivalent container types is IMO the most important consideration. 

CC @lawnjelly @hpvb It would be great for you two to weigh in with your thoughts sometime soon so we can merge this and begin the process of migrating over to Span

**Comment by Repiteo** [comment]
`size_t` wouldn't change based on bitcount, you're thinking of `uintptr_t`

**Comment by akien-mga** [comment]
> `size_t` wouldn't change based on bitcount, you're thinking of `uintptr_t`

https://stackoverflow.com/a/918909

**Comment by lawnjelly** [comment]
> size_t wouldn't change based on bitcount, you're thinking of uintptr_t

Yes `size_t` is imo well worth avoiding in the codebase specifically because it is afaik architecture dependent and we are multiplatform. Most programmers have moved to fixed width types these days because they avoid a host of problems with the old style types (can change the size of structures, padding etc, implicit conversions). (Afaik there can be some small exceptions in optimized tight loops, but these are few and far between)

This is the same reason we prefer `int32_t` over `int` where possible.

**Comment by lawnjelly** [comment]
My thoughts here would be not to use `size_t` because non-fixed type.
I would argue for templating the size like we do in `LocalVector`.

Personally I would also default `U` to `uint32_t` like in `LocalVector` (some people like to use signed types for counters, but I won't get into that war here). `Span` is to be used heavily internally and we want to avoid any penalties on any architectures when converting to 32 bit (if loop counters are 32 bit).

If you default to 64 bit, you then may end up having to do this (or the equivalent cast, either explicit or implicit):
```
uint32_t count = my_span.size();
for (uint32_t n=0; n<count; n++)
{
...
}
```
As well as extra code, this now creates a bug on every use case - what happens if the size *IS* larger than 32 bit? Same problem occurs with a cast. Are you going to write error checking code in every loop? Or are you going to force every loop to 64 bit *just in case*?

You also potentially have to make all your structures 64 bit *just in case*, and it can spread in the codebase, like a Nan.

The actual real world cases for 64 bit in loops are few and far between (memory addresses, large videos, files, not a lot else), and imo in those cases it is better to use a 64 bit version explicitly (e.g. with the templated `U` create a `Span64` which can be used as a parameter for functions).

In general I have seen some opinions that 64 bit is *free* but afaik it isn't - quite apart from any CPU costs it also has the potential to bloat structures and data, on both 32 bit and 64 bit platforms.

**Comment by clayjohn** [comment]
@lawnjelly When I wrote my comment I was forgetting that LocalVector uses uint32_t. I thought that we had standardized on uint64_t for both LocalVector and Vector. Looking through the codebase it seems that LocalVector is templated and can take another type for its size, but it appears that, in practice, it always uses uint32_t. 

The only concern I have for templating is the edge case where spans are parameters to methods that can take either Vectors or LocalVectors. If Span is templated, then those methods need to be templated too, which will cause a lot of pain

**Comment by Repiteo** [comment]
> `size_t` is definitely architecture dependent: https://stackoverflow.com/a/918909

I'll be damned! I always interpreted the specification as equivalent to a max possible size, but maybe I crossed some wires with `uintmax_t`.

In any case, I'm not totally against `size_t`. Despite being a fringe type in theory, it remains **the** type used for virtually all native size operations in C/C++. Granted, I would extend this point to say that we should be using `size_t` in virtually all of these cases, but that's well beyond the scope of this PR and probably isn't a hill worth dying on.

It's also worth noting that there already was a transition to 64-bit for CowData in #86730, which plenty of container types build on top of. It's implemented in a somewhat archaic manner with `Size`/`USize` typedefs, but the core is `uint64_t`. If we want Span to be "just" 32-bit, it would be worth looking into the problems that were initially solved by the 64-bit conversion. But this might be more of a topic for a broader examination of the different sizes we use across templates, let alone the codebase.

**Comment by lawnjelly** [comment]
> The only concern I have for templating is the edge case where spans are parameters to methods that can take either Vectors or LocalVectors.

Just trying to guess what you mean here, do you mean if we default to `u32` and e.g. `Vector` is trying to pass into a larger than `u32` size to a `Span32` (just calling it that here for disambiguation)?

We could just check this on the span constructor that it's within the 32 bit range.

For functions that are likely to need 64 bit input just make them take `Span64` as a parameter instead of `Span32`. Job done. `Span64` could be created automatically from `Vector` or `LocalVector` just like `Span32`, but doesn't need the size check on construction.

It is possible on much modern hardware there may be less penalties from 64 bit in terms of CPU so maybe we could get away with it, but bear in mind even the size of counters is doubled, stack size, storage size etc. I come from the school of making use of every byte so it's hard to adjust to, especially with how rarely 64 bit is actually required. :grin:

Remember 32 bit is enough to access 4Gb of `byte` size units.
16Gb of `uint32_t` units.
48Gb of `Vector3(float)` units.
96Gb of `Vector3(double)` units.
Outside of specialist areas like files and video, those kind of sizes are going to run so slow you should be using chunked access anyway. The cache will likely be dying. They aren't going to be super useful for games imo. Which is why I would treat 64 bit as special case.

We presumably also have a lot of existing code that uses 32 bit counters, and if swapping to 64 bit we should probably also consider swapping loops to e.g.:
```
for (uint64_t n=0; n<span.size(); n++)
```
otherwise each of these would be an overflow potential, and we should also evaluate whether there are implicit conversion costs when comparing 32 to 64 bit.

So in summary, I personally think 64 bit counters throughout core is madness, but go for it if you're feeling lucky. :wink: 

User code is another matter. Bound languages are going to be so slow the 32/64 bit makes no odds, so I'm not surprised it was used in bound containers.

**Comment by YYF233333** [comment]
Some thoughts: I think we could store a `size_t` value internally in `Span` while exposing an `int64_t` interface externally. The advantage of this approach is that `Span` itself will always take up exactly two machine words without padding. Additionally, the size returned to the caller can be directly compared with `int` without warnings (which is an issue I often encounter when replacing `Vector` with `LocalVector`). This means existing loops using `int` as an index will continue to compile without modification. While, in theory, arrays larger than `INT64_MAX` could exist, in practice, no machine has that much memory yet.

File: core/templates/span.h [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef SPAN_H
+#define SPAN_H
```
Comment (YYF233333) [inline core/templates/span.h:32]: I think we should start to use `#pragma once` as #102298 is already merged?

Comment (Repiteo) [reply core/templates/span.h:32]: Correct; though this would be rebased before merging anyway from the `size_t` discussion, so it'll be addressed at that time

**Comment by Ivorforce** [comment]
Alright, thanks for all the perspectives!
I'll just add a few quick words of my own, then summarize where we are.

For me, `uint32_t` is a hard no. Practically, it means we'd be declaring now: Godot `Vector` will never be able to hold more than 4GB of raw buffer data. I think that's a pretty hard sell.

As for @lawnjelly's templates: I think that doesn't really solve any problems, because either 32-bit or 64-bit would have to be the default, and be used for `Vector` and the likes. Defaulting to 64 bit with an option to use 32 bit instead makes little sense to me. But defaulting to 32 bit data means that GDScript users won't ever be able to make > 4GB buffers. I think that's a hard sell.

> If you default to 64 bit, you then may end up having to do this (or the equivalent cast, either explicit or implicit): [...]

Generally, the speed of `uint32_t` and `uint64_t` are practically the same in most situations. I think differences like in looped conversions are unlikely to be a bottleneck. But my main argument here is that there's little reason to iterate with `uint32_t` in the first place, if we go with 64-bit indices.

### Summary (for now)

### `size_t`
- Fast and small on both 32 bit and 64 bit
- System library equivalent of sizes
- Discrepancy between 32 bit and 64 bit systems; may need special handling
- Not currently in use for any containers

### `uint32_t`
- Limited to 4GB byte arrays (more for other data types)
- Good on 32 bit (rare), mostly wasteful on 64 bit (common)
- `int64_t` must be used for indexing (otherwise 2GB max)
- `LocalVector` use this right now (by default)

### `uint64_t`
- Good on 64 bit (common), wasteful on 32 bit (rare)
- `CowData` / `Vector` / `String` use this right now


## Conclusion

Taking everyone's opinion into account, the least problematic contender seems to be `uint64_t`, and we can add some sanity checks on 32 bit systems to make sure we don't accidentally try to create spans that don't even fit in memory.

I'll adjust the PR accordingly. Feel free to approve the PR based on this change, or reject if you don't think the discussion has sufficiently covered all grounds (especially looking at @lawnjelly). We can always make the final decision in the next core meeting :)

**Review by clayjohn — APPROVED** [review]
Looks good on my end!

**Comment by Repiteo** [comment]
I'm open to revisiting `size_t` in the future, but it should be as a dedicated proposal/PR. Otherwise, my previous approval stands

**Review by lawnjelly — APPROVED** [review]
Barring my preference for 32 bit, it looks great, but I can see I'm outvoted on that and it's not a hill to die on.

As long as I get bragging rights when you later realise 32 bit was a better choice. :+1: :wink:

**Comment by Repiteo** [comment]
Thanks!
